tests/device_tests/test_msg_signtx.py

```diff
@@ -221,7 +221,86 @@ def test_testnet_fee_too_high(self, client):
             tx_hash(serialized_tx).hex()
-            == "c669527e0d80dc645925f6965e1622e71fa5ca51e284442c4620c1ade7a76c63"
+            == "54fd5e9b65b8acc10144c1e78ea9720df7606d7d4a543e4c547ecd45b2ae226b"
+        )
+
+    @pytest.mark.skip_t1
+    def test_testnet_fee_high_hardfail(self, client):
+        # tx: 6f90f3c7cbec2258b0971056ef3fe34128dbde30daa9c0639a898f9977299d54
+        # input 1: 10.00000000 BTC
+        inp1 = proto.TxInputType(
+            address_n=parse_path("44'/1'/0'/0/0"),
+            # amount=1000000000,
+            prev_hash=TXHASH_6f90f3,
+            prev_index=1,
+        )
+
+        out1 = proto.TxOutputType(
+            address="mfiGQVPcRcaEvQPYDErR34DcCovtxYvUUV",
+            amount=1000000000 - 500000000 - 80000000,
+            script_type=proto.OutputScriptType.PAYTOADDRESS,
+        )
+
+        out2 = proto.TxOutputType(
+            address_n=parse_path("44'/1'/0'/1/0"),
+            amount=500000000,
+            script_type=proto.OutputScriptType.PAYTOADDRESS,
+        )
+
+        with client:
+            client.set_expected_responses(
+                [
+                    request_input(0),
+                    request_meta(TXHASH_6f90f3),
+                    request_input(0, TXHASH_6f90f3),
+                    request_input(1, TXHASH_6f90f3),
+                    request_output(0, TXHASH_6f90f3),
+                    request_output(1, TXHASH_6f90f3),
+                    request_output(0),
+                    proto.ButtonRequest(code=B.ConfirmOutput),
+                    request_output(1),
+                    proto.Failure(code=proto.FailureType.DataError),
+                ]
+            )
+            with pytest.raises(TrezorFailure, match="Fee exceeds"):
+                btc.sign_tx(
+                    client, "Testnet", [inp1], [out1, out2], prev_txes=TX_CACHE_TESTNET
+                )
+
+        # set SafetyCheckLevel to Prompt and try again
+        with client:
+            client.set_expected_responses(
+                [
+                    proto.ButtonRequest(),
+                    proto.Success(),
+                    proto.Features(),
+                    request_input(0),
+                    request_meta(TXHASH_6f90f3),
+                    request_input(0, TXHASH_6f90f3),
+                    request_input(1, TXHASH_6f90f3),
+                    request_output(0, TXHASH_6f90f3),
+                    request_output(1, TXHASH_6f90f3),
+                    request_output(0),
+                    proto.ButtonRequest(code=B.ConfirmOutput),
+                    request_output(1),
+                    proto.ButtonRequest(code=B.FeeOverThreshold),
+                    proto.ButtonRequest(code=B.SignTx),
+                    request_input(0),
+                    request_output(0),
+                    request_output(1),
+                    request_output(0),
+                    request_output(1),
+                    request_finished(),
+                ]
+            )
```

Comment:
@matejcik is there any added value in testing all the expected responses in this test case? It makes sense mostly because of "B.FeeOverThreshold" but other than that it is almost annoying because we need to change a lot of tests when we change something. Maybe it would be interesting to introduce `client.assertItemInExpectedResponses(proto.ButtonRequest(code=B.FeeOverThreshold))`?

Also @mmilata please replace the "proto" with "messages", we are trying to get rid of that.

Reply:
that's easy enough to implement -- just record responses and then `assert something in client.responses`

maybe a better choice is to use the layout functionality though. Unfortunately that only works on TT, but this buttonrequest doesn't happen on T1 anyway.

Speaking of, I'm wondering what to do about the layouts thing when the port is done. Presumably some of the text will be wildly different. But maybe the extended ButtonRequest functionality will make up for it?

Reply:
Changed the test to check that it received the appropriate layout in 426f9511daff5023c186d12ea89ea2823d5526c5. Don't like that solution very much, thinking of something more general that can be reused by other tests.

Replaced `proto` with `messages` in fd7606fcf2cc0c063beae801660c2dea909b1bf6.

Reply:
As we have `set_input_flow` and `set_expected_responses`, I didn't want to add another mechanism for monitoring messages. Instead I tried to use the former in f45099d3. Thoughts?

Reply:
it's interesting, but not worth it IMHO. there are three button requests total; better write out all three layout assertions explicitly

Reply:
Changed to `set_input_flow` with explicit ButtonRequest assertions in 52b4502.